Aaron routes

AaronTheBruce opened the pull request:
Completed the POST route for calendars.
Logic checks to see if the user's requested start and end dates coincide with an existing date range for a given location's calendar schedule.

AaronTheBruce committed:
Testing Calendar Get Routes to_dictionary

AaronTheBruce committed:
Merge branch 'master' of https://github.com/chaunceykm/campy-backend into aaron-routes

AaronTheBruce committed:
Add GET to Calendar Routes

AaronTheBruce committed:
Resolve date access error for first half of for loop

AaronTheBruce committed:
Update Calendar Post route

AaronTheBruce committed:
Clean up calendar.py

AaronTheBruce committed:
Additional cleanup

AaronTheBruce committed:
Add Calendar PUT and DELETE routes

arsentieva commented on app/routes/calendar.py:
```diff
@@ -0,0 +1,60 @@
+                {
+                    "start_date": fields.Date(required=True, description="Calendar start date"),
+                    "end_date": fields.Date(required=True, description="Calendar end date"),
+                    "location_id": fields.Integer(required=True, description="Calendar location"),
```
can you please add` example = 1` for  location_id and user_id? It can be done on a separate change

AaronTheBruce replied:
Ok, I'm on it

arsentieva commented on app/routes/calendar.py:
```diff
@@ -0,0 +1,60 @@
+    def post(self, location_id):
+        '''Create a new calendar booking with the provided date range'''
+        data = api.payload
+        dates = Calendar.query.filter_by(location_id=location_id).first()
```
`dates = Calendar.query.filter_by(location_id=location_id).all()` in stead of .first()

AaronTheBruce replied:
Implemented requested change

arsentieva commented on app/routes/calendar.py:
```diff
@@ -0,0 +1,60 @@
+            return {"Message": "Successfully scheduled!"}
+        else:
+            # get the calendar entries for a given location
+            locationDates = Calendar.query.filter_by(location_id=location_id).all()
```
`locationDates = Calendar.query.filter_by(location_id=location_id).all()` looks like we are reading the db again

AaronTheBruce replied:
I made the switch to .all(), but since we are getting only 1 location. I believe we could also use .first() for the Calendar query.

arsentieva replied:
lets talk about this use case, i might be missing something

AaronTheBruce replied:
Per out verbal conversation I have removed the duplicate database call

AaronTheBruce committed:
Fix Typo

fixed class CalendarById typo on line 64

arsentieva commented on app/routes/calendar.py:
```diff
@@ -0,0 +1,60 @@
+            # get the api payload data for the requested start and end date and convert to datetime from string
+            req_start_date = datetime.datetime.strptime(data["start_date"], "%Y-%m-%d").date()
+            req_end_date = datetime.datetime.strptime(data["end_date"], "%Y-%m-%d").date()
+
```
we need also to check for the cases where then end date cannot be before the start date

AaronTheBruce replied:
I thought about that, but then I forgot to implement it because I tunnel visioned in on making the overlap checked. I'll add that. Thanks for the reminder!

arsentieva reviewed (changes requested):
please see comments below, feel free to ask for clarifications

AaronTheBruce committed:
Add example to Calendar models for location_id and user_id

AaronTheBruce committed:
Merge branch 'aaron-routes' of https://github.com/chaunceykm/campy-backend into aaron-routes

AaronTheBruce committed:
Line 45 change .first() to .all() for Calendar.query

AaronTheBruce committed:
Add logic to POST route that does not allow the user to select an end date before their start date

AaronTheBruce committed:
Move conditional logic for start and end date order before the for loop

AaronTheBruce committed:
Add error status code of 400 to line 50

AaronTheBruce committed:
Update POST and PUT routes to handle location max_days and edge cases

AaronTheBruce committed:
Merge branch 'aaron-routes' of https://github.com/chaunceykm/campy-backend into aaron-routes

AaronTheBruce commented:
I updated the POST and PUT routes.
- Implemented a check for max_days and number of days scheduled
- Check that the start date is before the end date
- Created a helper function to handle checking for schedule overlap

arsentieva commented on app/routes/calendar.py:
```diff
@@ -30,69 +31,123 @@ def post(self, location_id):
+        # get the specifc calendar date to edit
+        calendar_range = Calendar.query.get(int(id))
+        # get all the dates for the object to update on
+        database_dates = Calendar.query.filter_by(location_id=location_id).all()
```
what is the use case here? Aren't we already getting the calendar record on line 84?

AaronTheBruce replied:
calendar_ranges is the individual calendar range that we want to update
database_dates are all of the dates for the given location that we want to check against to make sure that the update does not overlap with a pre-existing calendar database entry.

arsentieva commented on app/routes/calendar.py:
```diff
@@ -30,69 +31,123 @@ def post(self, location_id):
+            return {"message": "Calendar Booking not found, nothing deleted"}, 404
+
+def check_max_days(location_id):
+    location = Location.query.filter_by(id=location_id).one()
```
Why do we need to read the database for a location record?

arsentieva replied:
Take this question back :)

arsentieva reviewed (commented):
Please see the comment.

ryankris82 reviewed (approved):
Wow! Looks great!
